Plugin.pm

```diff
@@ -954,31 +959,74 @@ sub QobuzGetTracks {
 					image => $formattedTrack->{image},
 					tracks => []
 				} unless $works->{$workId};
-
+				
+				if ($workId ne $lastwork) {  # create a new work heading
+#					if ($i && $lastwork eq "") {  # create a separator line after previous single track#
+#						push @$items,{
+#							name  => "--------",
+#							type  => 'text'
+#						};
+#					}
+					push @$items,{
+						name  => $work,
+						type  => 'text'
+					};
+						
+					$lastwork = $workId;
+				}
+				
 				push @{$works->{$workId}->{tracks}}, $formattedTrack;
-			}
+				if (scalar @{$works->{$workId}->{tracks}} > 1) {
+					$worksfound = 1;   # we found a work with more than one track
+				}
+			} elsif ($lastwork ne "") {  # create a separator line for tracks without a work
+						push @$items,{
+							name  => "--------",
+							type  => 'text'
+						};
+						
+					$lastwork = "";
+			}	
 
 			$i++;
 
 			push @$items, $formattedTrack;
 		}
 
 		if (scalar keys %$works) {
-			foreach my $work (sort { $works->{$b}->{index} <=> $works->{$a}->{index} } keys %$works) {
-				my $workTracks = $works->{$work}->{tracks};
-
-				# insert works item before the first of its tracks
-				splice @$items, $works->{$work}->{index}, 0, {
-					name => $works->{$work}->{title},
-					image => $works->{$work}->{image},
-					type => 'playlist',
-					playall => 1,
-					url => \&QobuzWorkGetTracks,
-					passthrough => [{
-						tracks => $workTracks
-					}],
-					items => $workTracks
-				} if scalar @$workTracks > 1;
+			if ( $prefs->get('workPlaylistPosition') eq "before" ) {
+				# insert work playlists before the tracks unless there is only one work containing all tracks
+				foreach my $work (sort { $works->{$b}->{index} <=> $works->{$a}->{index} } keys %$works) {
+					my $workTracks = $works->{$work}->{tracks};
+					splice @$items, 0, 0, {
```

Comment:
replace with `unshift @$items, {...}`

Reply:
I was unhappy with the whole "if (scalar keys %$works) {" block of code, it was there before, but had become a lot more complex after SamY's and then my changes, and it ended up with duplicated code because it was adding elements to the items array one at a time, which meant the splice/unshift had to be done with a reverse sort of the elements to be added. 

Your comments have pushed me to refactor it. I know I have introduced an intermediate array, but I think the code is now easier to read/maintain. What do you think?

```
		if (scalar keys %$works) {
			# create work playlists unless there is only one work containing all tracks
			my @workPlaylists = ();
			foreach my $work (sort { $works->{$a}->{index} <=> $works->{$b}->{index} } keys %$works) {
				my $workTracks = $works->{$work}->{tracks};
				if ( $worksfound && scalar @$workTracks && scalar @$workTracks < $album->{tracks_count} ) {
					push @workPlaylists, {
						name => $works->{$work}->{title},
						image => 'html/images/playlists.png',
						type => 'playlist',
						playall => 1,
						url => \&QobuzWorkGetTracks,
						passthrough => [{
							tracks => $workTracks
						}],
						items => $workTracks
					}
				}
			}
			if ( @workPlaylists ) {
				# insert work playlists according to the user preference
				if ( $prefs->get('workPlaylistPosition') eq "before" ) {
					unshift @$items, @workPlaylists;
				} elsif ( $prefs->get('workPlaylistPosition') eq "after" ) {
					push @$items, @workPlaylists;
				}
			}
		}

```

Reply:
Looks good to me. I was mostly commenting on the `splice @a, 0, 0, {}` which is technically the same as an unshift, but less readable.